polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -7,43 +7,71 @@ use polars_time::prelude::*;
 use super::*;
 
 #[cfg(feature = "date_offset")]
-pub(super) fn date_offset(s: Series, offset: Duration) -> PolarsResult<Series> {
+pub(super) fn date_offset(s: &[Series]) -> PolarsResult<Series> {
+    let sa = &s[0];
+    let sb = &s[1];
+
     let preserve_sortedness: bool;
-    let out = match s.dtype().clone() {
+    let out = match sa.dtype().clone() {
         DataType::Date => {
-            let s = s
+            let sa = sa
                 .cast(&DataType::Datetime(TimeUnit::Milliseconds, None))
                 .unwrap();
             preserve_sortedness = true;
-            date_offset(s, offset).and_then(|s| s.cast(&DataType::Date))
+            date_offset(&[sa, sb.clone()]).and_then(|s| s.cast(&DataType::Date))
         }
         DataType::Datetime(tu, tz) => {
-            let ca = s.datetime().unwrap();
+            let ca = sa.datetime().unwrap();
+            let offsets = sb.utf8().unwrap();
 
-            fn offset_fn(tu: TimeUnit) -> fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64> {
-                match tu {
-                    TimeUnit::Nanoseconds => Duration::add_ns,
-                    TimeUnit::Microseconds => Duration::add_us,
-                    TimeUnit::Milliseconds => Duration::add_ms,
-                }
-            }
+            let offset_fn = match tu {
+                TimeUnit::Nanoseconds => Duration::add_ns,
+                TimeUnit::Microseconds => Duration::add_us,
+                TimeUnit::Milliseconds => Duration::add_ms,
+            };
 
-            let out = match tz {
+            let tz_args = match tz {
                 #[cfg(feature = "timezones")]
-                Some(ref tz) => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, tz.parse::<Tz>().ok().as_ref()))
+                Some(ref tz) => tz.parse::<Tz>().ok(),
+                _ => None,
+            };
+
+            let out = match offsets.len() {
+                1 => {
+                    let offset = match offsets.get(0) {
+                        Some(offset) => Duration::parse(offset),
+                        _ => Duration::new(0),
+                    };
```

Comment:
design-wise, is the idea that if an element of `offset` is `None`, then the datetime column stays as-is?

arguably, should the `None` propagate, and the result be `None`? do you know what's done for other operations? (I don't know if there's a general rule)

Reply:
Yeah, that was the idea, I had seen `sum` treat nulls as `0` - so I followed that thought process.

There doesn't seem to be a general rule, and as you say, perhaps the `None` should propagate instead - I see #10016 is open to discuss this type of thing in a more general sense.

Reply:
ah that's a sum across the elements of a series

I think here it would be more natural to follow what happens when summing two Series, i.e. that the `None` propagates:
```python
In [16]: ser1 = pl.Series([1, 2, 3])

In [17]: ser2 = pl.Series([1, None, 1])

In [18]: ser1 + ser2
Out[18]:
shape: (3,)
Series: '' [i64]
[
        2
        null
        4
]
```
?

Reply:
Yep, thanks for the example - it makes sense to follow that behaviour.

Reply:
I should have just checked what happens with `pl.duration`:
```python
df = pl.select(date = pl.repeat(pl.date(2023, 7, 23), 3)).with_columns(days = pl.lit([1, None, 3]))

# shape: (3, 2)
# ┌────────────┬──────┐
# │ date       ┆ days │
# │ ---        ┆ ---  │
# │ date       ┆ i64  │
# ╞════════════╪══════╡
# │ 2023-07-23 ┆ 1    │
# │ 2023-07-23 ┆ null │
# │ 2023-07-23 ┆ 3    │
# └────────────┴──────┘

df.with_columns(new_date = pl.col('date') + pl.duration(days='days'))

# shape: (3, 3)
# ┌────────────┬──────┬────────────┐
# │ date       ┆ days ┆ new_date   │
# │ ---        ┆ ---  ┆ ---        │
# │ date       ┆ i64  ┆ date       │
# ╞════════════╪══════╪════════════╡
# │ 2023-07-23 ┆ 1    ┆ 2023-07-24 │
# │ 2023-07-23 ┆ null ┆ null       │
# │ 2023-07-23 ┆ 3    ┆ 2023-07-26 │
# └────────────┴──────┴────────────┘
```
Basically it seems we just want to create a `duration` type by running each non-null string through `Duration::parse()`
```python
s = pl.Series(['1d', '-20d', None, '1mo'])

# shape: (4,)
# Series: '' [str]
# [
# 	"1d"
# 	"-20d"
# 	null
# 	"1mo"
# ]
```
I'm a bit unsure on what the simplest way to do that is.

With the current approach of using `iter().zip()` and `Int64Chunked::from_vec()` - it seems like I'm at the wrong level to handle `None` - because I need to produce an `i64` value?

`pl.duration` itself seems to work at a different level, as it's dealing with `Expr` objects and using `Expr::AnonymousFunction`

https://github.com/pola-rs/polars/blob/56fd7849469bf73b14c7cd9057225e1d7649fde1/polars/polars-lazy/polars-plan/src/dsl/functions/temporal.rs#L280

Reply:
I just tried
```Rust
let out = datetime
    .into_iter()
    .zip(offsets.into_iter())
    .map(|(v, offset)| {
        match offset {
            Some(offset) => {
                let offset = Duration::parse(offset);
                Some(offset_fn(&offset, v.unwrap(), tz_args.as_ref()).unwrap())
            }
            _ => None
        }
    });
Ok(Int64Chunked::from_iter_options("", out))
```

and it seems to work

The `unwrap`s inside the `map` need taking care of though:
- `offset_by` might error (e.g. adding `'1mo'` to the 31st of January without `'_saturating'`, or adding one day to `2020-10-24 02:00 BST`
- `v` might be `None`

Reply:
This seems to work for removing the `unwrap`s:
```Rust
let out = datetime
    .into_iter()
    .zip(offsets.into_iter())
    .map(|(v, offset)| match (v, offset) {
        (Some(v), Some(offset)) => {
            let offset = Duration::parse(offset);
            Ok(Some(offset_fn(&offset, v, tz_args.as_ref()).unwrap()))
        }
        _ => Ok(None),
    })
    .collect::<PolarsResult<Vec<_>>>()?;
Ok(Int64Chunked::from_slice_options("", &out))
```

In any case, could you please expand the test cases to cover the above?
- when the date / datetime series contains missing values
- when `offset` contains missing values
- when adding the offset would result in an error (check, using `pytest.raises`, that an appropriate warning is raised, rather than a `panic`)

Reply:
Ah okay, `from_*_options()` is a thing.

Will work on the test cases - thanks for the tips.